compiler/util/OwnershipFlowManager.cpp

```diff
@@ -788,11 +784,73 @@ void OwnershipFlowManager::autoCopyForSimpleAssignment(CallExpr* call,
 
     processBitwiseCopy(lhse, prod, live, cons);
   }
+}
+
+// If the RHS is a call to a function, no autocopy is needed because calls
+// are uniformly treated as returning an owned value.  If the thing being
+// copied into the RVV is owned, we don't need to insert an autoCopy.
+// But that still leaves the case where a CallExpr on the RHS returns
+// something that is unowned.
```

Comment:
Remember that this effort is not about whether the code is correct, since it’s
almost certain that it’s not in toto, but do my refactoring steps correctly preserve
the current behavior.  Once we can read it, we can make better decisions about
which parts are correct and which parts need to be modified.





On Sep 18, 2015, at 12:58 PM, Lydia Duncan <notifications@github.com<mailto:notifications@github.com>> wrote:


In compiler/util/OwnershipFlowManager.cpp<https://github.com/chapel-lang/chapel/pull/2565#discussion_r39894664>:

> @@ -788,11 +784,73 @@ void OwnershipFlowManager::autoCopyForSimpleAssignment(CallExpr* call,
>
>      processBitwiseCopy(lhse, prod, live, cons);
>    }
> +}
> +
> +// If the RHS is a call to a function, no autocopy is needed because calls
> +// are uniformly treated as returning an owned value.  If the thing being
> +// copied into the RVV is owned, we don't need to insert an autoCopy.
> +// But that still leaves the case where a CallExpr on the RHS returns
> +// something that is unowned.


I'm not as confident on my grasp of when autocopy calls are and aren't required, so I feel hesitant to sign off on this justification. I would like @mppf<https://github.com/mppf> to verify that what this comment describes is true, but other than that the code looks fine.

—
Reply to this email directly or view it on GitHub<https://github.com/chapel-lang/chapel/pull/2565/files#r39894664>.